k8s/client.go

```diff
@@ -62,46 +89,198 @@ func New(apiServerURL string, caCert []byte, token string) (Client, error) {
 }
 
 func (c *client) GetIngresses() ([]Ingress, error) {
-	endpoint := c.baseURL + "/apis/extensions/v1beta1/ingresses"
-	req, err := http.NewRequest("GET", endpoint, nil)
+	var ingressList IngressList
+	err := c.unmarshallRequest(ingressPath, &ingressList)
 	if err != nil {
 		return nil, err
 	}
-	req.Header.Add("Authorization", "Bearer "+c.token)
+	return ingressList.Items, nil
+}
 
-	log.Debugf("k8s<-: %v", *req)
+func (c *client) WatchIngresses(w Watcher) error {
+	log.Debug("Adding watcher for ingresses")
 
-	resp, err := c.http.Do(req)
+	go func() {
+		ingressRequest := func() (*http.Response, error) {
+			resourceVersion, err := c.getIngressResourceVersion()
+			if err != nil {
+				return nil, err
+			}
+			log.Debugf("Found ingress resource version '%s'", resourceVersion)
+			return c.request(ingressPath + "?watch=true&resourceVersion=" + resourceVersion)
+		}
+
+		retryRequest := func() (*http.Response, error) {
+			return retryRequest(w, ingressRequest)
+		}
+
+		for watch(w, retryRequest) {
+		}
+
+		log.Debug("Watch has stopped")
+	}()
+
+	return nil
+}
+
+type genericList struct {
+	ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+}
+
+func (c *client) getIngressResourceVersion() (string, error) {
+	var ingresses genericList
+	err := c.unmarshallRequest(ingressPath, &ingresses)
 	if err != nil {
-		return nil, err
+		return "", err
+	}
+	return ingresses.ResourceVersion, nil
+}
+
+// watch returns true if it should be retried, false if the watcher has terminated.
+func watch(w Watcher, request func() (*http.Response, error)) bool {
+	resp, err := request()
+	if err != nil {
+		log.Infof("Watcher could not make request, shutting down: %v", err)
+		return false
 	}
 	defer resp.Body.Close()
+	log.Infof("Watching %v", resp.Request.URL)
 
-	if resp.StatusCode < 200 || 300 <= resp.StatusCode {
-		return nil, fmt.Errorf("GET %s returned %v", endpoint, resp)
+	// send an update for a successful watch start
+	w.Updates() <- struct{}{}
+
+	updateCh := make(chan interface{})
+
+	go func() {
+		defer close(updateCh)
+
+		scanner := bufio.NewScanner(resp.Body)
+
+		for scanner.Scan() {
+			event := scanner.Text()
+			log.Infof("Received event %s", event)
+			updateCh <- struct{}{}
+		}
+
+		if err := scanner.Err(); err != nil {
+			log.Debugf("Error while watching events, closing update channel: %v", err)
+		}
+	}()
+
+	for {
+		var reconnect bool
+
+		select {
+		case <-w.Done():
+			log.Debug("Watcher is done, stopping watch")
+			return false
+		case update := <-updateCh:
+			if update == nil {
+				log.Info("Long poll terminated, will reconnect")
+				reconnect = true
```

Comment:
Does the break come out of the fall loop? If so could this return there instead of breaking?